src/runtime/storageNG/reference-mode-store.ts

```diff
@@ -0,0 +1,598 @@
+        const {pendingIds, model} = this.constructPendingIdsAndModel(this.containerStore.localModel.getData());
+        const send = () => void this.callbacks.get(message.id)({type: ProxyMessageType.ModelUpdate, model: model(), id: message.id});
+        if (pendingIds.length === 0) {
+          this.enqueueSend(send);
+        } else {
+          this.enqueueBlockingSend(pendingIds, send);
+        }
+        
+        break;
+      }
+      default:
+          throw new Error('Unexpected ProxyMessageType');
+    }
+    return true;
+  }
+
+  /**
+   * Enqueues a sending function on the send queue. If the send queue is empty then
+   * the function is immediately invoked.
+   */
+  private enqueueSend(runnable: Runnable) {
+    if (this.pendingSends.length === 0) {
+      runnable();
+    } else {
+      this.pendingSends.push({fn: runnable});
+    }
+  }
+
+  /**
+   * Enqueues a send function on the send queue, deferring execution until the
+   * provided id list is available in the backing store.
+   */
+  private enqueueBlockingSend(ids: string[], runnable: Runnable) {
+    const block = Math.random() + '';
+    this.pendingSends.push({fn: runnable, block});
+    this.holdQueue.enqueue(ids, () => this.processPendingSends(block));
+  }
+
+  /**
+   * Process any sends in the pending send queue, including sends blocked on the
+   * provided block. This should only be called by the holdQueue.
+   */
+  private processPendingSends(block: string) {
+    while (this.pendingSends.length > 0) {
+      if (this.pendingSends[0].block == null || this.pendingSends[0].block === block) {
+        const send = this.pendingSends.shift();
+        send.fn();
+      }
+    }
+  } 
+
+  /**
+   * Convert the provided entity to a CRDT Model of the entity. This requires synthesizing
+   * a version map for the CRDT model, which is also provided as an output.
+   */
+  private entityToModel(entity: Entity): {model: EntityData<S, C>, version: VersionMap} {
+    if (this.versions[entity.id] == undefined) {
+      this.versions[entity.id] = {};
+    }
+    const entityVersion = this.versions[entity.id];
+    const model = new this.backingStore.modelConstructor().getData();
+    let maxVersion = 0;
+    for (const key of Object.keys(entity)) {
+      if (model.singletons[key]) {
+        if (entityVersion[key] == undefined) {
+          entityVersion[key] = 0;
+        }
+        const version = {[this.crdtKey]: ++entityVersion[key]};
+        maxVersion = Math.max(maxVersion, entityVersion[key]);
+        model.singletons[key].values = {[entity[key].id]: {value: entity[key], version}};
+        model.singletons[key].version = version;
+      } else if (model.collections[key]) {
+        model.collections[key].values = {};
+        if (entityVersion[key] == undefined) {
+          entityVersion[key] = 0;
+        }
+        const version = {[this.crdtKey]: ++entityVersion[key]};
+        maxVersion = Math.max(maxVersion, entityVersion[key]);
```

Comment:
Lines 405-409 looks duplicated from the if-statement above. Could you hoist them outside the if-statements?

You might have an issue when `key === id` (you don't want to bump versions for that), but that is probably better dealt with explicitly anyway by adding `if (key === id) continue;` at the top of the for loop -- that seems clearer to me.

Reply:
Yup! Moving maxVersion updating outside the if/then works.